openedx/core/djangoapps/monitoring_utils/utils.py

```diff
@@ -0,0 +1,99 @@
+from django.conf import settings
+
+indices = {}
+
+# The directory in which graph files will be created.
+GRAPH_DIRECTORY_PATH = settings.MEMORY_GRAPH_DIRECTORY
+
+# The max number of object types for which to show data on the console
+MAX_CONSOLE_ROWS = 10
+
+# The max number of object types for which to generate reference graphs
+MAX_GRAPHED_OBJECT_TYPES = 5
+
+# Maximum depth of forward reference graphs
+REFS_DEPTH = 3
+
+# Maximum depth of backward reference graphs
+BACK_REFS_DEPTH = 8
+
+# Max number of objects per type to use as starting points in the reference graphs
+MAX_OBJECTS_PER_TYPE = 5
+
+# Object type names for which graphs should not be generated even if the new
+# object count is high.  "set" is ignored by default because many sets are
+# created in the course of tracking the number of new objects of each type.
+IGNORED_TYPES = ('set',)
+
+log = logging.getLogger(__name__)
+
+
+def show_memory_leaks(
+        label=u'memory_leaks',
+        max_console_rows=MAX_CONSOLE_ROWS,
+        max_graphed_object_types=MAX_GRAPHED_OBJECT_TYPES,
+        refs_depth=REFS_DEPTH,
+        back_refs_depth=BACK_REFS_DEPTH,
+        max_objects_per_type=MAX_OBJECTS_PER_TYPE,
+        ignored_types=IGNORED_TYPES,
+        graph_directory_path=GRAPH_DIRECTORY_PATH):
+    """
+    Call this function to get data about memory leaks; what objects are being
+    leaked, where did they come from, and what do they contain?  The leaks
+    are measured from the last call to ``objgraph.get_new_ids()`` (which is
+    called within this function).  Some data is printed to stdout, and more
+    details are available in graphs stored at the paths printed to stdout.
+    Subsequent calls with the same label are indicated by an increasing index
+    in the filename.
+
+    Args:
+        label (unicode): The start of the filename for each graph
+        max_console_rows (int): The max number of object types for which to
+            show data on the console
+        max_graphed_object_types (int): The max number of object types for
+            which to generate reference graphs
+        refs_depth (int): Maximum depth of forward reference graphs
+        back_refs_depth (int): Maximum depth of backward reference graphs
+        max_objects_per_type (int): Max number of objects per type to use as
+            starting points in the reference graphs
+        ignored_types (iterable): Object type names for which graphs should
+            not be generated even if the new object count is high.
+        graph_directory_path(unicode): The directory in which graph files
+            will be created.  It will be created if it doesn't already exist.
+    """
+    new_objects_output = StringIO()
+    new_ids = objgraph.get_new_ids(limit=max_console_rows, file=new_objects_output)
+    log.info('\n' + new_objects_output.getvalue())
+    index = indices.setdefault(label, 1)
+    indices[label] = index + 1  # Set the index for the next call with the same label
```

Comment:
I don't know if it's any more intuitive, but when I've had per-label counters before, one thing I've done is to use `defaultdict` with a default value of `itertools.count`. To advance the counter and get the value, you do a `next` on the dict.